<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Проверим-нет-ли-в-данных-пропусков-или-аномалий" data-toc-modified-id="Проверим-нет-ли-в-данных-пропусков-или-аномалий-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Проверим нет ли в данных пропусков или аномалий</a></span></li><li><span><a href="#Раздление-признаков" data-toc-modified-id="Раздление-признаков-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Раздление признаков</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Модель-случайного-леса" data-toc-modified-id="Модель-случайного-леса-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Модель случайного леса</a></span><ul class="toc-item"><li><span><a href="#Модель-случайного-леса-(поиск-во-вложенном-цикле)" data-toc-modified-id="Модель-случайного-леса-(поиск-во-вложенном-цикле)-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Модель случайного леса (поиск во вложенном цикле)</a></span></li></ul></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Сокращение-выборки-(Downsampling)" data-toc-modified-id="Сокращение-выборки-(Downsampling)-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Сокращение выборки (Downsampling)</a></span><ul class="toc-item"><li><span><a href="#Модель-случайного-леса" data-toc-modified-id="Модель-случайного-леса-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Модель случайного леса</a></span></li><li><span><a href="#Модель-логистической-регрессии" data-toc-modified-id="Модель-логистической-регрессии-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Модель логистической регрессии</a></span></li></ul></li><li><span><a href="#Увеличение-выборки-(Upsampling)" data-toc-modified-id="Увеличение-выборки-(Upsampling)-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Увеличение выборки (Upsampling)</a></span><ul class="toc-item"><li><span><a href="#Модель-случайного-леса-(вложенный-цикл)" data-toc-modified-id="Модель-случайного-леса-(вложенный-цикл)-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Модель случайного леса (вложенный цикл)</a></span></li><li><span><a href="#Модель-случайного-леса-(2-одиночных-цикла)" data-toc-modified-id="Модель-случайного-леса-(2-одиночных-цикла)-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Модель случайного леса (2 одиночных цикла)</a></span></li><li><span><a href="#Модель-логистической-регрессии" data-toc-modified-id="Модель-логистической-регрессии-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Модель логистической регрессии</a></span></li></ul></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

In [40]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.utils import shuffle
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import f1_score

from sklearn.preprocessing import StandardScaler 

import math



<div style="border:solid purple 5px; padding: 20px"> 
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>

<h3> Широкоформатный Jupyter <a class="tocSkip"> </h3>

Расширяем границы, или как сделать работу более комфортной (не всем нравится 😄)

    from IPython.core.display import display, HTML
    display(HTML("<style>.container { width:90% !important; }</style>"))

In [41]:
data = pd.read_csv('/datasets/Churn.csv')  #.sort_index(axis=0, ascending=True).head(100)

In [42]:
# Посмотрим структуру данных
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [43]:
# оценим даные визуально
data.head(11)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0           1    15634602  Hargrave          619    France  Female   42   
1           2    15647311      Hill          608     Spain  Female   41   
2           3    15619304      Onio          502    France  Female   42   
3           4    15701354      Boni          699    France  Female   39   
4           5    15737888  Mitchell          850     Spain  Female   43   
5           6    15574012       Chu          645     Spain    Male   44   
6           7    15592531  Bartlett          822    France    Male   50   
7           8    15656148    Obinna          376   Germany  Female   29   
8           9    15792365        He          501    France    Male   44   
9          10    15592389        H?          684    France    Male   27   
10         11    15767821    Bearce          528    France    Male   31   

    Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0      2.0       0.00              1          1               1   
1      1.0   83807.86              1          0               1   
2      8.0  159660.80              3          1               0   
3      1.0       0.00              2          0               0   
4      2.0  125510.82              1          1               1   
5      8.0  113755.78              2          1               0   
6      7.0       0.00              2          1               1   
7      4.0  115046.74              4          1               0   
8      4.0  142051.07              2          0               1   
9      2.0  134603.88              1          1               1   
10     6.0  102016.72              2          0               0   

    EstimatedSalary  Exited  
0         101348.88       1  
1         112542.58       0  
2         113931.57       1  
3          93826.63       0  
4          79084.10       0  
5         149756.71       1  
6          10062.80       0  
7         119346.88       1  
8          74940.50       0  
9          71725.73       0  
10         80181.12       0

### Проверим нет ли в данных пропусков или аномалий

In [44]:
# проверим нет ли пропусков или аномалий в "значимых" строковых признаках

print(data.value_counts(['Geography']))
print(data.value_counts(['Gender']))
print('\n ..пропусков или аномалий нет.')

Geography
France       5014
Germany      2509
Spain        2477
dtype: int64
Gender
Male      5457
Female    4543
dtype: int64

 ..пропусков или аномалий нет.


In [45]:
# проверим нет ли пропусков или аномалий в "значимых" бинарных признаках

print(data.value_counts(['Exited']))
print(data.value_counts(['NumOfProducts']))
print(data.value_counts(['HasCrCard']))
print(data.value_counts(['IsActiveMember']))
print('\n ..пропусков или аномалий нет.')


Exited
0         7963
1         2037
dtype: int64
NumOfProducts
1                5084
2                4590
3                 266
4                  60
dtype: int64
HasCrCard
1            7055
0            2945
dtype: int64
IsActiveMember
1                 5151
0                 4849
dtype: int64

 ..пропусков или аномалий нет.


In [46]:
# проверим нет ли аномальных значений в "значимых" вещественных признаках
print(data['Balance'].min(), data['Balance'].max())
print(data['EstimatedSalary'].min(), data['EstimatedSalary'].max())
#print(data.value_counts(['Tenure']))
print(data['Tenure'].isna().count())
print('\n ..аномалий нет.')

0.0 250898.09
11.58 199992.48
10000

 ..аномалий нет.


In [47]:
print(data.loc[data['Tenure'].isna(),:]['RowNumber'].count())
print('\n столько строк имеют значение Nan в атрибуте Tenure.')

909

 столько строк имеют значение Nan в атрибуте Tenure.


***Пропуски значений в атрибуте Tenure заменим на 0.0 т.к. этот призна тоже участвует в обучении*** 

<div class="alert alert-info"> <b>Комментарий студента:</b> Круто, спасибо за описаные способы!</div>

In [48]:

data['Tenure'] = data['Tenure'].fillna(0.0)
print(data.loc[data['Tenure'].isna(),:]['RowNumber'].count())

0


 ### Раздление признаков

In [49]:
# Проведем дамми-нормализацию по тем строковым атрибутам, которые имеют потенциальное влияние на выходной параметр.
# Таковыми являются только 'Geography','Gender'
# Атрибуты 'RowNumber','CustomerId','Surname' не имеют никакого "бизнес"-смысла - это суррогатные ключи, 
# либо как 'Surname' - это просто фамилия клиента.Считаем , что она не является влияющим фактором.

data_dummies = pd.get_dummies(data, columns=['Geography','Gender'], drop_first=True)#, drop_first=True

data_train, data_test = train_test_split(data_dummies, test_size=0.2, random_state=12345)
data_train, data_valid = train_test_split(data_train, test_size=0.25, random_state=12345)

print(data_train.shape)
print(data_valid.shape)
print(data_test.shape)
print('')

data_train.info()

(6000, 15)
(2000, 15)
(2000, 15)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6000 entries, 492 to 3295
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   RowNumber          6000 non-null   int64  
 1   CustomerId         6000 non-null   int64  
 2   Surname            6000 non-null   object 
 3   CreditScore        6000 non-null   int64  
 4   Age                6000 non-null   int64  
 5   Tenure             6000 non-null   float64
 6   Balance            6000 non-null   float64
 7   NumOfProducts      6000 non-null   int64  
 8   HasCrCard          6000 non-null   int64  
 9   IsActiveMember     6000 non-null   int64  
 10  EstimatedSalary    6000 non-null   float64
 11  Exited             6000 non-null   int64  
 12  Geography_Germany  6000 non-null   uint8  
 13  Geography_Spain    6000 non-null   uint8  
 14  Gender_Male        6000 non-null   uint8  
dtypes: float64(3), int64(8), object(1), 

In [50]:
pd.options.mode.chained_assignment = None

# Опишем список колонок для стандартизации. Стандартизовать нужно только численные признаки
#numeric_col = ['CreditScore','Age','Tenure','Balance', 'NumOfProducts'
#                 ,'HasCrCard','IsActiveMember','EstimatedSalary'
#                ,'Geography_Germany','Geography_Spain','Gender_Male']

numeric_col = ['CreditScore','Age','Tenure','Balance', 'NumOfProducts','EstimatedSalary' ]


In [51]:
# разобъем признаки на обучающую выборку, валидационную и тестовую

target_test = data_test['Exited']
features_test = data_test.drop(['RowNumber','CustomerId','Surname','Exited'], axis =1)

target_valid = data_valid['Exited']
features_valid = data_valid.drop(['RowNumber','CustomerId','Surname','Exited'], axis =1)

target_train = data_train['Exited']
features_train = data_train.drop(['RowNumber','CustomerId','Surname','Exited'], axis =1)

# выведем информацию о структуре датафрейма
print(features_train.info())


<class 'pandas.core.frame.DataFrame'>
Int64Index: 6000 entries, 492 to 3295
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CreditScore        6000 non-null   int64  
 1   Age                6000 non-null   int64  
 2   Tenure             6000 non-null   float64
 3   Balance            6000 non-null   float64
 4   NumOfProducts      6000 non-null   int64  
 5   HasCrCard          6000 non-null   int64  
 6   IsActiveMember     6000 non-null   int64  
 7   EstimatedSalary    6000 non-null   float64
 8   Geography_Germany  6000 non-null   uint8  
 9   Geography_Spain    6000 non-null   uint8  
 10  Gender_Male        6000 non-null   uint8  
dtypes: float64(3), int64(5), uint8(3)
memory usage: 439.5 KB
None


In [52]:
scaler = StandardScaler()
scaler.fit(features_train[numeric_col]) 
features_train[numeric_col] = scaler.transform(features_train[numeric_col])

#scaler.fit(features_valid[numeric_col])
features_valid[numeric_col] = scaler.transform(features_valid[numeric_col])

#scaler.fit(features_test[numeric_col])
features_test[numeric_col] = scaler.transform(features_test[numeric_col])



In [53]:
# Выведер размеры выборок
print('\nРазмеры выборок:')
print('Обучающая: ',features_train.shape)
print('Валидирующая: ',features_valid.shape)
print('Тестовая: ',features_test.shape)

print('Длина обучающих выходных признаков', target_train.shape) 
print('Длина валидирующих выходных признаков', target_valid.shape)


Размеры выборок:
Обучающая:  (6000, 11)
Валидирующая:  (2000, 11)
Тестовая:  (2000, 11)
Длина обучающих выходных признаков (6000,)
Длина валидирующих выходных признаков (2000,)


In [54]:
# визуально оценим результат
#print(features_train.head(5))
display(features_train.head(7))
display(target_train.head(7))
#display(target_train.value_counts())

display(features_valid.head(7))
display(target_valid.head(7))

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
492     -0.134048 -0.078068 -0.174142  0.076163       0.816929          0   
6655    -1.010798  0.494555 -1.451074  0.136391      -0.896909          1   
4287     0.639554  1.353490 -1.131841  0.358435      -0.896909          1   
42      -0.990168  2.116987 -0.812608  0.651725      -0.896909          1   
8178     0.567351  0.685430  0.783558  0.813110       0.816929          1   
2127     0.598296 -0.459816  1.422024  0.921963       0.816929          1   
2051     0.618925 -0.173505 -1.131841  0.482629       0.816929          0   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
492                1         0.331571                  0                0   
6655               1        -0.727858                  0                0   
4287               1        -0.477006                  1                0   
42                 1        -0.100232                  0                0   
8178               1         0.801922                  0                0   
2127               0         0.820416                  1                0   
2051               0         1.208574                  0                0   

      Gender_Male  
492             0  
6655            1  
4287            1  
42              0  
8178            0  
2127            1  
2051            0

492     0
6655    0
4287    1
42      0
8178    0
2127    0
2051    0
Name: Exited, dtype: int64

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
2358     0.175393  0.399118 -1.131841  1.385698      -0.896909          0   
8463    -1.299609  0.971741 -0.812608 -1.232442      -0.896909          1   
163      0.711757 -0.268942 -0.812608 -1.232442       0.816929          1   
3074    -0.391916  0.494555  0.464325  0.672529      -0.896909          1   
5989     0.165078  1.353490  1.741257  0.536522      -0.896909          0   
6230    -0.618839  3.071359 -0.174142 -1.232442       0.816929          0   
6973     0.072246  0.017370  1.102791  0.472650       0.816929          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
2358               1        -1.466761                  0                0   
8463               0         0.254415                  0                1   
163                1         0.122863                  0                1   
3074               0         0.585847                  1                0   
5989               0         1.462457                  0                0   
6230               1        -1.442219                  0                0   
6973               1         1.708147                  1                0   

      Gender_Male  
2358            1  
8463            1  
163             0  
3074            0  
5989            0  
6230            1  
6973            1

2358    0
8463    1
163     0
3074    1
5989    1
6230    0
6973    0
Name: Exited, dtype: int64

In [55]:
# Оценим, насколько разбалансированы значения выходных параметров

print(target_train.value_counts())
#

0    4781
1    1219
Name: Exited, dtype: int64


## Исследование задачи

In [56]:
# оценим баланс классов на всей выборке:

print('Кол-во признаков соответствующих классу 0:', data_dummies[data_dummies['Exited']==0]['Exited'].count())
print('Кол-во признаков соответствующих классу 1:', data_dummies[data_dummies['Exited']==1]['Exited'].count())

Кол-во признаков соответствующих классу 0: 7963
Кол-во признаков соответствующих классу 1: 2037


***Вывод: Выборка достаточно сильно разбалансирована***

### Модель случайного леса

In [57]:
# Найдем параметры модели "случайного леса" с наилучшими результатами.
# Сначала найдем налучшую максимальную глубину

#best_rfc_model = None
#best_rfc_f1 = 0.0
#best_rfc_est = 0
#best_rfc_depth = 1

#for depth in range (2, 20):
   
#    model_rfc = RandomForestClassifier(random_state=12345, max_depth=depth)
#    model_rfc.fit(features_train, target_train) 
    # предсказание модели на обучающей выборке
#    predictions_train = model_rfc.predict(features_train)
    # посчитаем значение метрики F1 на обучающей выборке
#    rfc_result_tr_f1 = f1_score(target_train, predictions_train)

    # предсказание модели на валидационной выборке
#    predictions_valid = model_rfc.predict(features_valid)
    # посчитаем значение метрики F1 на валидационной выборке
#    rfc_result_v_f1 = f1_score(target_valid, predictions_valid)    
    
#    print('глубина:', depth)    
#    print('f1 (train):', rfc_result_tr_f1, end='')    
#    print('   f1 (valid):', rfc_result_v_f1)
#    if rfc_result_v_f1 > best_rfc_f1:
#        best_rfc_model = model_rfc
#        best_rfc_f1 = rfc_result_v_f1
#        best_rfc_depth = depth
       
        
#print('\nRandomForestClassifier:')            
#print('Лучшая максимальная глубина:',best_rfc_depth)
#print('F1 наилучшей модели (train):',best_rfc_f1)


In [58]:
# затем найдем кол-во деревьев дающее наилучшую оценку для F1

#best_rfc_model = None
#best_rfc_f1 = 0.000001
#best_rfc_est = 0


#for est in range (10, 110, 5):

#    model_rfc = RandomForestClassifier(random_state=12345, max_depth = best_rfc_depth, n_estimators = est)
#    model_rfc.fit(features_train, target_train) 

    # предсказание модели на обучающей выборке
    #predictions_train = model_rfc.predict(features_train)
    # посчитаем значение метрики F1 на обучающей выборке
    #rfc_result_tr_f1 = f1_score(target_train, predictions_train)

    # предсказание модели на валидационной выборке
#    predictions_valid = model_rfc.predict(features_valid)
    # посчитаем значение метрики F1 на валидационной выборке
#    rfc_result_v_f1 = f1_score(target_valid, predictions_valid)

#    print( np.around(rfc_result_v_f1,decimals=4), end = ' ')    
#    if rfc_result_v_f1 > best_rfc_f1:
#        best_rfc_model = model_rfc
#        best_rfc_f1 = rfc_result_v_f1
#        best_rfc_est = est
#        best_rfc_depth = depth
#print('')
#print(predictions_train)
#print(target_train)
#print(rfc_result_tr_f1)

#print('\nRandomForestClassifier:')            
#print('Лучшее кол-во деревьев:',best_rfc_est)
#print('Лучшая глубина:', best_rfc_depth)
#print('F1 наилучшей модели:',best_rfc_f1) 


####  Модель случайного леса (поиск во вложенном цикле)

In [59]:
# вложенный цикл с проходом по обоим параметрам
#найдем кол-во деревьев дающее наилучшую оценку для F1

best_rfc_model = None
best_rfc_f1 = 0.000001
best_rfc_est = 0
best_rfc_depth = 2

for depth in range (2, 17):
#    print('depth:', depth)
#    print('f1:',end=' ')
    #print(' F1 наилучшей модели:',best_rfc_f1)
    for est in range (30, 120, 5):

        model_rfc = RandomForestClassifier(random_state=12345, max_depth = depth, n_estimators = est)
        model_rfc.fit(features_train, target_train) 

        # предсказание модели на валидационной выборке
        predictions_valid = model_rfc.predict(features_valid)
        # посчитаем значение метрики F1 на валидационной выборке
        rfc_result_v_f1 = f1_score(target_valid, predictions_valid)

        #print( np.around(rfc_result_v_f1,decimals=4), end = ' ')    
        if rfc_result_v_f1 > best_rfc_f1:
            best_rfc_model = model_rfc
            best_rfc_f1 = rfc_result_v_f1
            best_rfc_est = est
            best_rfc_depth = depth
    #print('')

print('\nRandomForestClassifier (вложеный цикл):')            
print('Лучшее кол-во деревьев:',best_rfc_est)
print('Лучшая глубина:', best_rfc_depth)
print('F1 наилучшей модели:',best_rfc_f1) 


RandomForestClassifier (вложеный цикл):
Лучшее кол-во деревьев: 40
Лучшая глубина: 13
F1 наилучшей модели: 0.5691318327974277


***Вывод: на несбалансированных данных значение F1 получилось достаточно маленькое. Скорее всего диспропорции к размерах выборок принадлижащих разным классам имеют большое влияниеи.*** 

***Сбалансируем данные, проверим..***

### Логистическая регрессия

In [60]:
best_lr_model = None
best_lr_f1 = 0.0
best_lr_solver = ''
best_lr_trsh = 0.0

solvers = ['liblinear','lbfgs','newton-cg', 'sag', 'saga']

for solver_ in solvers:
    model_lr = LogisticRegression(random_state=12345, solver = solver_)
    model_lr.fit(features_train, target_train)

    probabilities_valid = model_lr.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    
    print(solver_)
#    print(probabilities_one_valid)
    
    for threshold in np.arange(0, 0.6, 0.025):
        predicted_valid = list(map((lambda x: x > threshold),probabilities_one_valid)) 
#        predicted_valid = probabilities_one_valid > threshold
        f1 = f1_score(target_valid, predicted_valid)
#        print('threshold: ',threshold,'  f1:',f1)
        if f1 > best_lr_f1:
            best_lr_f1 = f1
            best_lr_trsh = threshold
            best_lr_model = model_lr
            best_lr_solver = solver_

print('\nBest solver:', best_lr_solver)
print('Best best_lr_trsh: ', np.around(best_lr_trsh,decimals=3))
print('Best F1 value: ', best_lr_f1)


liblinear
lbfgs
newton-cg
sag
saga

Best solver: liblinear
Best best_lr_trsh:  0.325
Best F1 value:  0.49343832020997375


## Борьба с дисбалансом

### Сокращение выборки (Downsampling)

In [61]:
print('Исходное кол-во признаков с группах 0 и 1:')
print(data_train.value_counts(['Exited']))

ones_count = data_train.loc[data_train['Exited']==1,:]['Exited'].count()

data_train_dwn_0 = data_train.loc[data_train['Exited']==0,:].sample(ones_count)
data_train_dwn_1 = data_train.loc[data_train['Exited']==1,:]

data_balanced_dwn = pd.concat([data_train_dwn_0]+[data_train_dwn_1])
print('')
print('Размерность сбалансированной выборки:',data_balanced_dwn.shape)
print('')
print('Новое кол-во признаков с группах 0 и 1:')
print(data_balanced_dwn.value_counts(['Exited']))
print('')
data_balanced_dwn.info()

Исходное кол-во признаков с группах 0 и 1:
Exited
0         4781
1         1219
dtype: int64

Размерность сбалансированной выборки: (2438, 15)

Новое кол-во признаков с группах 0 и 1:
Exited
0         1219
1         1219
dtype: int64

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2438 entries, 5193 to 7661
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   RowNumber          2438 non-null   int64  
 1   CustomerId         2438 non-null   int64  
 2   Surname            2438 non-null   object 
 3   CreditScore        2438 non-null   int64  
 4   Age                2438 non-null   int64  
 5   Tenure             2438 non-null   float64
 6   Balance            2438 non-null   float64
 7   NumOfProducts      2438 non-null   int64  
 8   HasCrCard          2438 non-null   int64  
 9   IsActiveMember     2438 non-null   int64  
 10  EstimatedSalary    2438 non-null   float64
 11  Exited             2438 no

In [62]:
# разобъем сбалансированный DF на признаки и выходные параметры + уберем неинформативные признаки
target_train_dwn = data_balanced_dwn['Exited']
features_train_dwn = data_balanced_dwn.drop(['RowNumber','CustomerId','Surname','Exited'], axis =1)

#  посмотрим размерности получившихся выборок
print(features_train_dwn.shape)
print(target_train_dwn.shape)


(2438, 11)
(2438,)


In [63]:
# Проведем дамми-нормализацию для сбалансированных наборов

#data_balanced = pd.concat([data_0]+[data_1])

#data_dummies = pd.get_dummies(data_balanced, columns=['Geography','Gender'])#, drop_first=True

# Выделим входные и выходные признаки
#target = data_balanced['Exited']
#features = data_dummies.drop(['RowNumber','CustomerId','Surname','Exited'], axis =1)

# Опишем список колонок для стандартизации. Стандартизовать нужно только численные признаки
#numeric_col = ['CreditScore','Age','Tenure','Balance', 'NumOfProducts','HasCrCard','IsActiveMember','EstimatedSalary']
#numeric_col = ['CreditScore','Age','Tenure','Balance', 'NumOfProducts','HasCrCard','IsActiveMember','EstimatedSalary']

#print(data_balanced.shape)

In [64]:
#приведем значения входных признаков к одному масштабу. Переиспользуем scaler из п 1.2 

features_train_dwn[numeric_col] = scaler.transform(features_train_dwn[numeric_col])
display(features_train_dwn)

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
5193    -0.928280 -0.937002  1.422024  0.698070      -0.896909          1   
4926     0.226966 -0.268942  0.464325  0.384443      -0.896909          1   
5107     2.011410 -0.173505  1.422024  0.576191      -0.896909          1   
2328     0.557037 -0.937002 -0.812608 -1.232442       0.816929          1   
4485     1.021198 -0.459816  1.422024  0.508813       0.816929          1   
...           ...       ...       ...       ...            ...        ...   
4654    -0.763245 -0.746128  0.145091  0.405721       4.244606          0   
8991    -1.083001  0.303681 -1.451074  1.057444      -0.896909          1   
4913    -0.928280  1.162616 -1.131841  0.767952      -0.896909          1   
3238     0.237281  1.448927  0.464325 -1.232442      -0.896909          1   
7661    -0.412545  0.494555 -0.812608  0.615524      -0.896909          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
5193               1        -1.619977                  0                1   
4926               0         1.430731                  1                0   
5107               0         0.281512                  0                1   
2328               1        -1.420612                  0                1   
4485               1        -1.266785                  1                0   
...              ...              ...                ...              ...   
4654               0         1.530055                  1                0   
8991               0        -0.238472                  1                0   
4913               1        -0.600620                  0                0   
3238               0         0.180203                  0                0   
7661               1        -0.038931                  1                0   

      Gender_Male  
5193            0  
4926            0  
5107            0  
2328            1  
4485            1  
...           ...  
4654            0  
8991            0  
4913            0  
3238            1  
7661            1  

[2438 rows x 11 columns]

In [65]:
#  визуально оценим промежуточные преобразованя
display(features_train_dwn.head(10))
display(target_train_dwn.head(10))
#  проверим что выборки теперь сбанасированны
display(target_train_dwn.value_counts())

display(features_valid.head(10))
display(target_valid.head(10))

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
5193    -0.928280 -0.937002  1.422024  0.698070      -0.896909          1   
4926     0.226966 -0.268942  0.464325  0.384443      -0.896909          1   
5107     2.011410 -0.173505  1.422024  0.576191      -0.896909          1   
2328     0.557037 -0.937002 -0.812608 -1.232442       0.816929          1   
4485     1.021198 -0.459816  1.422024  0.508813       0.816929          1   
2390    -0.515692 -0.650691  0.783558  1.731596       0.816929          1   
614      0.371372 -1.223314  1.102791  1.112484      -0.896909          1   
845     -1.516218  0.971741  0.783558  0.051296      -0.896909          0   
7052    -0.484748 -0.650691  0.145091 -1.232442       0.816929          1   
8068    -0.340342 -0.937002 -1.451074  0.355451      -0.896909          0   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
5193               1        -1.619977                  0                1   
4926               0         1.430731                  1                0   
5107               0         0.281512                  0                1   
2328               1        -1.420612                  0                1   
4485               1        -1.266785                  1                0   
2390               0         0.241233                  1                0   
614                1         1.311388                  1                0   
845                0         0.613490                  0                0   
7052               1        -1.005183                  0                0   
8068               1         0.553165                  1                0   

      Gender_Male  
5193            0  
4926            0  
5107            0  
2328            1  
4485            1  
2390            1  
614             1  
845             1  
7052            0  
8068            1

5193    0
4926    0
5107    0
2328    0
4485    0
2390    0
614     0
845     0
7052    0
8068    0
Name: Exited, dtype: int64

0    1219
1    1219
Name: Exited, dtype: int64

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
2358     0.175393  0.399118 -1.131841  1.385698      -0.896909          0   
8463    -1.299609  0.971741 -0.812608 -1.232442      -0.896909          1   
163      0.711757 -0.268942 -0.812608 -1.232442       0.816929          1   
3074    -0.391916  0.494555  0.464325  0.672529      -0.896909          1   
5989     0.165078  1.353490  1.741257  0.536522      -0.896909          0   
6230    -0.618839  3.071359 -0.174142 -1.232442       0.816929          0   
6973     0.072246  0.017370  1.102791  0.472650       0.816929          1   
5313     1.351269  3.166796  0.145091  0.712646       0.816929          1   
1131    -1.567791 -0.173505  1.422024  0.786561      -0.896909          1   
3369     0.526093 -0.555253  1.102791  1.825481      -0.896909          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
2358               1        -1.466761                  0                0   
8463               0         0.254415                  0                1   
163                1         0.122863                  0                1   
3074               0         0.585847                  1                0   
5989               0         1.462457                  0                0   
6230               1        -1.442219                  0                0   
6973               1         1.708147                  1                0   
5313               1         0.092048                  0                0   
1131               0         0.204979                  0                1   
3369               1        -0.346512                  0                0   

      Gender_Male  
2358            1  
8463            1  
163             0  
3074            0  
5989            0  
6230            1  
6973            1  
5313            0  
1131            1  
3369            0

2358    0
8463    1
163     0
3074    1
5989    1
6230    0
6973    0
5313    0
1131    0
3369    0
Name: Exited, dtype: int64

In [66]:
# Выведем размеры выборок

print('\nРазмеры выборок:')
print('Обучающая (после downsampling): ',features_train_dwn.shape)
print('Валидирующая: ',features_valid.shape)
print('Тестовая: ',features_test.shape)

print('Длина обучающих выходных признаков', target_train_dwn.shape)


Размеры выборок:
Обучающая (после downsampling):  (2438, 11)
Валидирующая:  (2000, 11)
Тестовая:  (2000, 11)
Длина обучающих выходных признаков (2438,)


#### Модель случайного леса

In [67]:
# ищем модель с наилучшим кол-вом estimate'оров 

best_rfc_est = 0
best_rfc_depth = 2
best_rfc_f1 = 0.000
best_rfc_dwn_model = None

#print(features_train.shape)
for depth in range (5, 13):
    print('depth:', depth)
    print('f1:',end=' ')
    for est in range (40, 150, 5):

        model_rfc = RandomForestClassifier(random_state=12345, max_depth = depth, n_estimators = est)
        model_rfc.fit(features_train_dwn, target_train_dwn) 

        # предсказание модели на валидационной выборке
        predictions_valid = model_rfc.predict(features_valid)
        # посчитаем значение метрики F1 на валидационной выборке
        rfc_result_v_f1 = f1_score(target_valid, predictions_valid)

        print( np.around(rfc_result_v_f1,decimals=3), end = ' ') 
        if rfc_result_v_f1 > best_rfc_f1:
            best_rfc_dwn_model = model_rfc
            best_rfc_f1 = rfc_result_v_f1
            best_rfc_est = est
            best_rfc_depth = depth
    #print('prediction_valid:', predictions_valid) 
    #print('target_valid:', target_valid) 
    #print('f1:', rfc_result_v_f1) 
    print('')

print('\nRandomForestClassifier (Downsampling):')            
print('Лучшее кол-во деревьев:',best_rfc_est)
print('Лучшая глубина:', best_rfc_depth)
print('F1 наилучшей модели:',best_rfc_f1) 

depth: 2
f1: 0.514 0.526 0.508 0.509 0.513 0.511 0.52 0.523 0.514 0.516 0.517 0.515 0.513 0.516 0.517 0.535 0.529 0.515 0.517 0.52 0.52 0.523 0.532 0.531 
depth: 3
f1: 0.52 0.513 0.52 0.516 0.52 0.519 0.521 0.521 0.529 0.521 0.531 0.524 0.539 0.537 0.542 0.548 0.547 0.545 0.544 0.546 0.543 0.547 0.547 0.547 
depth: 4
f1: 0.542 0.542 0.545 0.545 0.544 0.544 0.544 0.539 0.543 0.543 0.548 0.546 0.542 0.545 0.545 0.544 0.543 0.543 0.54 0.542 0.542 0.541 0.541 0.537 
depth: 5
f1: 0.56 0.545 0.55 0.551 0.547 0.547 0.552 0.551 0.553 0.55 0.552 0.55 0.549 0.55 0.549 0.547 0.546 0.545 0.544 0.546 0.545 0.546 0.544 0.548 
depth: 6
f1: 0.553 0.548 0.555 0.558 0.561 0.566 0.566 0.561 0.568 0.565 0.563 0.553 0.553 0.559 0.559 0.56 0.555 0.561 0.56 0.56 0.564 0.564 0.563 0.568 
depth: 7
f1: 0.549 0.557 0.553 0.553 0.559 0.562 0.561 0.563 0.566 0.561 0.564 0.562 0.564 0.558 0.564 0.563 0.566 0.57 0.568 0.57 0.571 0.569 0.569 0.565 
depth: 8
f1: 0.563 0.551 0.554 0.555 0.563 0.567 0.567 0.569 0.576 0.

#### Модель логистической регрессии

In [68]:
best_lr_f1 = 0.0
best_lr_trsh = 0.0
best_lr_dwn_model = None

#print(features_train.shape)

model_lr = LogisticRegression(random_state=12345, solver = 'liblinear')
model_lr.fit(features_train_dwn, target_train_dwn)

probabilities_valid = model_lr.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
    
for threshold in np.arange(0, 0.6, 0.025):
    predicted_valid = list(map((lambda x: x > threshold),probabilities_one_valid)) 
    f1 = f1_score(target_valid, predicted_valid)
#    print('threshold: ',threshold,'  f1:',f1)
    if f1 > best_lr_f1:
        best_lr_f1 = f1
        best_lr_trsh = threshold
        best_lr_dwn_model = model_lr

print('\nBest best_lr_trsh: ', best_lr_trsh)
print('Best F1 value: ', best_lr_f1)


Best best_lr_trsh:  0.525
Best F1 value:  0.48971962616822434


***Вывод: балансировка групп признаков дала положительный результат.*** 

***Получено значение метрики F1:***

***~0.827988~ 0.5754 -для модели случайного леса,***

***~0.7896~ 0.4854 -для модели лог.регрессии.***

***- в обоих случаях ~больше~  ниже чем задано в постановке задачи.***

### Увеличение выборки (Upsampling)

In [69]:
#сбалансируем наборы признаков путем увеличение меньшей группы до размера бОльшей

multipl = math.floor(np.around(data_train.loc[data_train['Exited']==0,:]['Exited'].count() 
                        / data_train.loc[data_train['Exited']==1,:]['Exited'].count()
                    ,decimals=0))

data_train_up_0 = data_train.loc[data_train['Exited']==0,:]
data_train_up_1 = data_train.loc[data_train['Exited']==1,:]


print('Во сколько раз нужно увеличить кол-во признаков соответствующих меньшему классу :', multipl)
print('data_train_up_0:',data_train_up_0.shape)
print('data_train_up_1:',data_train_up_1.shape)

data_balanced_up = pd.concat([data_train_up_0]+[data_train_up_1]*multipl)

print('Размерность сбалансированной выборки:',data_balanced_up.shape)

data_balanced_up.info()

Во сколько раз нужно увеличить кол-во признаков соответствующих меньшему классу : 4
data_train_up_0: (4781, 15)
data_train_up_1: (1219, 15)
Размерность сбалансированной выборки: (9657, 15)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 9657 entries, 492 to 7661
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   RowNumber          9657 non-null   int64  
 1   CustomerId         9657 non-null   int64  
 2   Surname            9657 non-null   object 
 3   CreditScore        9657 non-null   int64  
 4   Age                9657 non-null   int64  
 5   Tenure             9657 non-null   float64
 6   Balance            9657 non-null   float64
 7   NumOfProducts      9657 non-null   int64  
 8   HasCrCard          9657 non-null   int64  
 9   IsActiveMember     9657 non-null   int64  
 10  EstimatedSalary    9657 non-null   float64
 11  Exited             9657 non-null   int64  
 12  Geography_Germany  9657 n

In [70]:
# разобъем на признаки и выходные параметры + уберем неинформативные признаки
target_train_up = data_balanced_up['Exited']
features_train_up = data_balanced_up.drop(['RowNumber','CustomerId','Surname','Exited'], axis =1)


print('Общий размер новой выборки:',data_balanced_up.shape)
print('Кол-во с Exited = 0:',data_balanced_up.loc[data_balanced_up['Exited']==0,:]['Exited'].count())
print('Кол-во с Exited = 1:',data_balanced_up.loc[data_balanced_up['Exited']==1,:]['Exited'].count())

Общий размер новой выборки: (9657, 15)
Кол-во с Exited = 0: 4781
Кол-во с Exited = 1: 4876


In [71]:
# делаем нормализацию

features_train_up[numeric_col] = scaler.transform(features_train_up[numeric_col])
display(features_train_up)

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
492     -0.134048 -0.078068 -0.174142  0.076163       0.816929          0   
6655    -1.010798  0.494555 -1.451074  0.136391      -0.896909          1   
42      -0.990168  2.116987 -0.812608  0.651725      -0.896909          1   
8178     0.567351  0.685430  0.783558  0.813110       0.816929          1   
2127     0.598296 -0.459816  1.422024  0.921963       0.816929          1   
...           ...       ...       ...       ...            ...        ...   
4654    -0.763245 -0.746128  0.145091  0.405721       4.244606          0   
8991    -1.083001  0.303681 -1.451074  1.057444      -0.896909          1   
4913    -0.928280  1.162616 -1.131841  0.767952      -0.896909          1   
3238     0.237281  1.448927  0.464325 -1.232442      -0.896909          1   
7661    -0.412545  0.494555 -0.812608  0.615524      -0.896909          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
492                1         0.331571                  0                0   
6655               1        -0.727858                  0                0   
42                 1        -0.100232                  0                0   
8178               1         0.801922                  0                0   
2127               0         0.820416                  1                0   
...              ...              ...                ...              ...   
4654               0         1.530055                  1                0   
8991               0        -0.238472                  1                0   
4913               1        -0.600620                  0                0   
3238               0         0.180203                  0                0   
7661               1        -0.038931                  1                0   

      Gender_Male  
492             0  
6655            1  
42              0  
8178            0  
2127            1  
...           ...  
4654            0  
8991            0  
4913            0  
3238            1  
7661            1  

[9657 rows x 11 columns]

In [72]:
# Выведем размеры выборок

print('\nРазмеры выборок:')
print('Обучающая (после upsampling): ',features_train_up.shape)
print('Валидирующая: ',features_valid.shape)
print('Тестовая: ',features_test.shape)

print('Длина обучающих выходных признаков', target_train_up.shape)


Размеры выборок:
Обучающая (после upsampling):  (9657, 11)
Валидирующая:  (2000, 11)
Тестовая:  (2000, 11)
Длина обучающих выходных признаков (9657,)


<div style="border:solid purple 5px; padding: 20px"> 
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>

<h3> Встроенные функции all() и any() <a class="tocSkip"> </h3>

Полезные функции, которые помогут узнать есть ли определенные значения в объекте.

`all` возвращает **True**, если все объекты равны **True** или удовлетворяют условию

![](https://i.ibb.co/NKs2yWq/image.png)

`any` в свою очередь делает то же самое, но возвращает **True**,если хотя бы один равен **True**

![](https://i.ibb.co/4tq8TmV/image.png)

#### Модель случайного леса (вложенный цикл)

In [73]:
best_rfc_f1 = 0.0
best_rfc_up_model = None
best_rfc_depth = 1

for depth in range (2, 17):
    print('depth - ', depth, end=' :')
    for est in range (10, 130, 5):
   
        model_rfc_up = RandomForestClassifier(random_state=12345, max_depth = depth, n_estimators = est)
        model_rfc_up.fit(features_train_up, target_train_up) 

        # предсказание модели на валидационной выборке
        predict_valid_up = model_rfc_up.predict(features_valid)
        # посчитаем значение метрики F1 на валидационной выборке
        rfc_result_v_up_f1 = f1_score(target_valid, predict_valid_up)
        print(est,end=' ')    
        #print('f1 (train):', rfc_result_tr_f1, end='')
        #print(' f1 (valid):', rfc_result_v_f1)    
        if rfc_result_v_up_f1 > best_rfc_f1:
            best_rfc_up_model = model_rfc_up
            best_rfc_f1 = rfc_result_v_up_f1
            best_rfc_est = est
            best_rfc_depth = depth
    print(', max F1: ',np.round(best_rfc_f1, decimals = 3))

print('\nRandomForestClassifier (Upsampling):')            
print('Лучшее кол-во деревьев:',best_rfc_est)
print('Лучшая глубина:', best_rfc_depth)
print('F1 наилучшей модели:',best_rfc_f1) 

depth -  2 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.527
depth -  3 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.54
depth -  4 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.554
depth -  5 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.556
depth -  6 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.568
depth -  7 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.58
depth -  8 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.593
depth -  9 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.597
depth -  10 :10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 , max F1:  0.597
depth -  11 :10 15 20 25 30 35 40 45 5

#### Модель случайного леса (2 одиночных цикла)

In [74]:
best_rfc_f1 = 0.0
best_rfc_up1_model = None
best_rfc_depth_1 = 1

for depth in range (2, 17):

    model_rfc_up1 = RandomForestClassifier(random_state=12345, max_depth = depth)
    model_rfc_up1.fit(features_train_up, target_train_up) 

    # предсказание модели на валидационной выборке
    predict_valid_up1 = model_rfc_up1.predict(features_valid)
    # посчитаем значение метрики F1 на валидационной выборке
    rfc_result_v_1_f1 = f1_score(target_valid, predict_valid_up1)
    print(depth, end='')    
    #print('f1 (train):', rfc_result_tr_f1, end='')
    #print(' f1 (valid):', rfc_result_v_f1)    
    if rfc_result_v_1_f1 > best_rfc_f1:
        best_rfc_up1_model = model_rfc_up1
        best_rfc_f1 = rfc_result_v_1_f1
        best_rfc_est = est
        best_rfc_depth_1 = depth
    print(', max F1: ',np.round(best_rfc_f1, decimals = 3))

print('\nRandomForestClassifier (Upsampling p1):')            
print('Лучшая глубина:', best_rfc_depth_1)
print('F1 наилучшей модели:',best_rfc_f1) 

2, max F1:  0.522
3, max F1:  0.539
4, max F1:  0.545
5, max F1:  0.545
6, max F1:  0.557
7, max F1:  0.575
8, max F1:  0.586
9, max F1:  0.586
10, max F1:  0.586
11, max F1:  0.586
12, max F1:  0.592
13, max F1:  0.592
14, max F1:  0.592
15, max F1:  0.592
16, max F1:  0.595

RandomForestClassifier (Upsampling p1):
Лучшая глубина: 16
F1 наилучшей модели: 0.5953991880920163


In [75]:
best_rfc_f1 = 0.0
best_rfc_up2_model = None
best_rfc_est = 1


for est in range (10, 130, 5):
   
    model_rfc_up2 = RandomForestClassifier(random_state=12345, max_depth = best_rfc_depth_1, n_estimators = est)
    model_rfc_up2.fit(features_train_up, target_train_up) 

    # предсказание модели на валидационной выборке
    predict_valid_up2 = model_rfc_up2.predict(features_valid)
    # посчитаем значение метрики F1 на валидационной выборке
    rfc_result_v_2_f1 = f1_score(target_valid, predict_valid_up2)
    print(est,end=' ')    
    #print('f1 (train):', rfc_result_tr_f1, end='')
    #print(' f1 (valid):', rfc_result_v_f1)    
    if rfc_result_v_2_f1 > best_rfc_f1:
        best_rfc_up2_model = model_rfc_up2
        best_rfc_f1 = rfc_result_v_2_f1
        best_rfc_est = est
    print(', max F1: ',np.round(best_rfc_f1, decimals = 3))

print('\nRandomForestClassifier (Upsampling):')            
print('Лучшее кол-во деревьев:',best_rfc_est)
print('F1 наилучшей модели:',best_rfc_f1) 

10 , max F1:  0.55
15 , max F1:  0.552
20 , max F1:  0.56
25 , max F1:  0.57
30 , max F1:  0.574
35 , max F1:  0.58
40 , max F1:  0.58
45 , max F1:  0.58
50 , max F1:  0.586
55 , max F1:  0.586
60 , max F1:  0.59
65 , max F1:  0.59
70 , max F1:  0.59
75 , max F1:  0.592
80 , max F1:  0.592
85 , max F1:  0.592
90 , max F1:  0.592
95 , max F1:  0.592
100 , max F1:  0.595
105 , max F1:  0.595
110 , max F1:  0.595
115 , max F1:  0.595
120 , max F1:  0.596
125 , max F1:  0.596

RandomForestClassifier (Upsampling):
Лучшее кол-во деревьев: 120
F1 наилучшей модели: 0.5959183673469387


***Вывод: ради эксперемента сравнил результаты работы вложенного цикла (с перебором depth и est) в п 3.2.1 и двух одиночных циклов с последовательным перебором depth, а затем est, в п3.2.2.  Результирующие значения F1 почти не отличаются! Можно ли как-то понять почему? Если результат не сильно отличается то смысла делать вложенный цикл нет, т.к. это пустая трата ресурсов и времени. Или я что-то не учел ?***

####  Модель логистической регрессии

In [76]:
best_lr_f1 = 0.0
best_lr_trsh = 0.0
best_lr_up_model = None

print(features_train.shape)

model_lr_up = LogisticRegression(random_state=12345, solver = 'liblinear')
model_lr_up.fit(features_train_up, target_train_up)

probabilities_valid = model_lr_up.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
    
for threshold in np.arange(0, 0.6, 0.025):
    predicted_valid = list(map((lambda x: x > threshold),probabilities_one_valid)) 
    f1 = f1_score(target_valid, predicted_valid)
#    print('threshold: ',threshold,'  f1:',f1)
    if f1 > best_lr_f1:
        best_lr_f1 = f1
        best_lr_trsh = threshold
        best_lr_up_model = model_lr_up

print('\nBest threshold: ', best_lr_trsh)
print('Best F1 value: ', best_lr_f1)

(6000, 11)

Best threshold:  0.525
Best F1 value:  0.485473289597001


***Вывод: лучший результат показала модель случайного леса, обученная после балансировки через Upsampling, с количеством деревьев 140 и максимальной глубиной одерева 17***

***Значение F1 = 0.9024 - выше чем при Downsamplin'е. Это и понятно, модель получилась более точной, т.к. мы максимально использовали все данные для обучения***

***Модель логистической регрессии обученная после балансировки через Upsampling, хоть и показала значение F1 выше порогового (0.59) но оказалась почему-то даже ниже чем в п.3.1.2***



## Тестирование модели

In [77]:
#Тестирование будем выполнять на случайного леса с макс.глубиной 17 и количеством деревьев 140, полученной после Upsampling
#

In [78]:
predictions_test = best_rfc_up_model.predict(features_test)

# посчитаем значение метрики F1 на тестовой выборке
rfc_test_f1 = f1_score(target_test, predictions_test)

print('F1 для тестовой выборки:', rfc_test_f1)
print(f1_score(target_test, best_rfc_up1_model.predict(features_test)))
print(f1_score(target_test, best_rfc_up2_model.predict(features_test)))

F1 для тестовой выборки: 0.6139747995418099
0.5953608247422679
0.5891472868217054


***Вывод: тест на наилучшей модели дал значение F1 - также, больше чем то, которое нужно было достич по заданию (0.59). Бинго!***



## Общий вывод



***1) Был проведен анализ исходных данных на предмет наличия пропусков и аномальных значений***

***2) Весь массив данных был поделен на 3 части : обучающая выборка, валидирующа, тестовая, в пропорции 3:1:1 соответственно. Проведена dummy-перекодировка для категориальных признаков, была выполнена нормализация всех наборов признаков через Scaler***

***В п.2 было обучены 2 модели : случайного леса и логистической регрессии. Для модели случайного леса подобраны оптимальные значения максимальной глубины деревьев и кол-ва естиматоров: 13 и 40 соотвественно. Достигнуто максимальное значение метрики F1 равное 0.5691. Значение прямо скажем не большое , но это скорее обусловлено дисбалансом кол-ва признаков относящихся к различным целевым классам Exited=0 и Exited=0. Модель логистической регрессии показала на данной обучающей выбьорке лучший результат при значении solver = liblinear, при этом значени F1 = 0.4934 - меньше чем у модели случ.леса***

***3) Проведена балансировка обучающей выборки методом downsampling , получены значения F1 0.5806 и 0.4868 для модели случ.леса и логистической регрессии соответственно. Результаты не совсем хорошие.***

***4) Проведена балансировка обучающей выборки методом upsampling , получены значения F1 0.6056 и 0.4854 для модели случ.леса и логистической регрессии соответственно. Модель случайного леса показала наилучший результат при значения гиперпараметров: максимальная глубина дерева 13 и число естиматоров 60***

***5) Тестирование на тестовой выборке было сделано на модели случайного леса с параметрами (макс.глубина = 12, кол-во эстиматоров = 60 - может варьироваться от запуска к запуску) при этом получено значение F1 равное 0.6139 - выше указаной в задании значения (0.59) -  значит считаем что цель работы достигнута !***

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.